### PR TITLE
GC: correctly handle manifest unknown (404) condition in v2DeleteManifest retry loop

### LINKED ISSUE #18381: GC retries forever on permanent errors, most commonly on "manifest unknown" (404)
**Expected behavior and actual behavior:**
When GC encounters a permanent error such as 404 when deleting Blobs or Manifests, it should not retry as there's no expectation that an object that was not found would appear if we retried.

GC Job log shows:
```
2023-03-20T18:37:49Z [INFO] [/jobservice/job/impl/gc/util.go:67]: failed to exec v2DeleteManifest, error: http status code: 404, body: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
, will retry again after: 171.764815ms
2023-03-20T18:37:50Z [INFO] [/jobservice/job/impl/gc/util.go:67]: failed to exec v2DeleteManifest, error: http status code: 404, body: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
, will retry again after: 152.306794ms
2023-03-20T18:37:50Z [INFO] [/jobservice/job/impl/gc/util.go:67]: failed to exec v2DeleteManifest, error: http status code: 404, body: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
, will retry again after: 623.007798ms
```

so it keeps retrying ad gets the same 404 every time, eventually it gives up, but since it's still in `artifact_thrash` it will go through the same process again next time.

This causes several issues:
1. when a 404 error occurs, the GC job enters a retry loop with an exponential backoff, each time getting the same 404 error back
2. this greatly increases the time it takes for this job to run
3. when it finally gives up, the thrash artifact that had the 404 error is not removed so it lingers in `artifact_thrash` table in the DB and when GC runs the next time the above is repeated all over again.

Since the goal of GC is to delete the referenced manifest or blob, if the response from the storage is that it in fact doesn't exist (404), then that's the desired end result anyway and shouldn't be treated as a retryable error, instead it should identify this condition, remove it from `artifact_thrash` and proceed.

**Steps to reproduce the problem:**
I don't have an exact repro as there could be different scenarios for why the blob is no longer in the backend storage of the registry, but on a given 404 error, the GC should should mark the manifest/blob is deleted, not retry and remove it from `artifact_thrash` table.

**Versions:**
Please specify the versions of following systems.

- harbor version: 2.7.1
- docker engine version: N/A
- docker-compose version: N/A

**Additional context:**

We use GCS as our registry storage backend.

GC job log will show a bunch of these for every "manifest unknown":
```
2023-03-20T18:37:49Z [INFO] [/jobservice/job/impl/gc/util.go:67]: failed to exec v2DeleteManifest, error: http status code: 404, body: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
, will retry again after: 171.764815ms
2023-03-20T18:37:50Z [INFO] [/jobservice/job/impl/gc/util.go:67]: failed to exec v2DeleteManifest, error: http status code: 404, body: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
, will retry again after: 152.306794ms
2023-03-20T18:37:50Z [INFO] [/jobservice/job/impl/gc/util.go:67]: failed to exec v2DeleteManifest, error: http status code: 404, body: {"errors":[{"code":"MANIFEST_UNKNOWN","message":"manifest unknown"}]}
, will retry again after: 623.007798ms
```

Example DELETE request to registry:
```
time="2023-03-20T19:09:45.893971097Z" level=info msg="authorized request" auth.user.name=harbor_registry_user go.version=go1.19.4 http.request.host="harbor-registry:5000" http.request.id=583cc81a-2fd8-46a0-8f97-a50762adf971 http.request.method=DELETE http.request.remoteaddr="172.26.6.225:49696" http.request.uri="/v2/<repo path>/manifests/sha256:dbaafd81b7cf24f1cff51402b7c1fdca832afbbc7280d7c52d1a23042408bf67" http.request.useragent=harbor-registry-client instance.id=2941b623-ec93-4e55-b9e4-06d833b41f29 service=registry vars.name=<repo> vars.reference="sha256:dbaafd81b7cf24f1cff51402b7c1fdca832afbbc7280d7c52d1a23042408bf67"```
```

Response with 404 from registry:
```
time="2023-03-20T19:09:45.925842324Z" level=error msg="response completed with error" auth.user.name=harbor_registry_user err.code="manifest unknown" err.message="manifest unknown" go.version=go1.19.4 http.request.host="harbor-registry:5000" http.request.id=583cc81a-2fd8-46a0-8f97-a50762adf971 http.request.method=DELETE http.request.remoteaddr="172.26.6.225:49696" http.request.uri="/v2/<repo path>/manifests/sha256:dbaafd81b7cf24f1cff51402b7c1fdca832afbbc7280d7c52d1a23042408bf67" http.request.useragent=harbor-registry-client http.response.contenttype=application/json http.response.duration=94.954574ms http.response.status=404 http.response.written=70 instance.id=2941b623-ec93-4e55-b9e4-06d833b41f29 service=registry vars.name=<repo> vars.reference="sha256:dbaafd81b7cf24f1cff51402b7c1fdca832afbbc7280d7c52d1a23042408bf67"
```

The above would repeat many times, and since this manifest is not removed from `artifact_thrash` the same will happen when GC runs again.

### PR DESCRIPTION
**Pull request by dkulchinsky**
# Comprehensive Summary of your change

Currently when a 404 error is returned from the registry in `v2DeleteManifest` retry loop, it will be considered is a temporary/retryable error and will eventually result in the `v2DeleteManifest` step to fail, although a 404 (MANIFEST_UNKNOWN) indicates that the manifest no longer exists.

This PR adds a check inside the retry loop that marks the `v2DeleteManifest` as success if registry returns a 404.

This change seem to correctly align with the next cleanup step that deletes all the manifests revisions using RegistryCtl which also treats a 404 error as success.

# Issue being fixed
Fixes #18381

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [x] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/jobservice/job/impl/gc/util.go [hunk 1]
```diff
@@ -62,6 +62,10 @@ func v2DeleteManifest(logger logger.Interface, repository, digest string) error
 		if err == readonly.Err {
 			return retry.Abort(err)
 		}
+		// If delete returned an err because the manifest is unknown, consider success
```
Comment (wy65701436) [inline src/jobservice/job/impl/gc/util.go:65]: the 404 is handled at https://github.com/goharbor/harbor/blob/main/src/jobservice/job/impl/gc/garbage_collection.go#L287

Comment (dkulchinsky) [reply src/jobservice/job/impl/gc/util.go:65]: it does, using a head request ince before entering the retry loop, however as I decribe in my comment here: https://github.com/goharbor/harbor/issues/18381#issuecomment-1477840214

you can see there is an edge case where the initial HEAD request returns 200, it enters the retry loop and tries to delete and gets a 500 at first and then on a subsequent retry gets a 404 which is treated as a retryable error (without the conditional I'm adding here) and eventually causes v2ManufesrDelete to return a not nil err.

this addresses this edge case.

**Comment by dkulchinsky** [comment]
Hey @Vad1mo, @wy65701436 was wondering if you had a chance to consider this PR?

I have a local build with this change running on our test Harbor and it is working very well, we're finally able to get GC running and actually cleaning up all these manifests & blobs.

**Review by zyyw — APPROVED** [review]

**Comment by dkulchinsky** [comment]
Thanks for approving @zyyw!

I updated the branch, let me know if there's anything else needed to get this merged.

**Comment by dkulchinsky** [comment]
Thanks @Vad1mo & @zyyw!

I wonder if it's possible to backport this change to v2.7? I realize v2.8 is almost out but we're not quite ready to upgrade yet as we're still working on our chartmuseum deprecation and would be great if we we could have this fix v2.7 for now

**Comment by Vad1mo** [comment]
Good question, would make sense as it is a fix. 
I would recommend to back port it yourself to 2.7 and open a PR for that

**Comment by Vad1mo** [comment]
you should also make one for the 2.8 release.
